### PR TITLE
fix: merge generation of input-data for `issue_closures` and `list_repos` 

### LINKED ISSUE #64: combine `issue_closures` and `list_repos` producer metrics into one producer
The `issue_closures` and `list_repos` metrics each produce data for every repository within some list of repository names (currently originating from the repos listed in `report.toml`. They each generate a `.csv` file in the generated `inputs` folder: 

- `metrics/list_repos.rs` gathers data that is ultimately generated in `$DATA_DIR/inputs/repo-infos.csv`. 

  That file might look like so: 
  
  ```csv
  # ,Repository Name ,# of PRs
  1 ,rust            ,452
  2 ,rustup          ,11
  3 ,rustfmt         ,12
  4 ,rustfix         ,0
  5 ,cargo           ,38
  6 ,crates.io       ,27
  7 ,docs.rs         ,4
  8 ,mdBook          ,9
  ```

- `metrics/issue_closure.rs` gathers data that is ultimately generated in `$DATA_DIR/inputs/issue-closure.csv`

  That file might look like so:
  
  ```csv
  # ,Organization ,Repository ,Issues Opened ,Issues Closed ,Start Date ,End Date
  1 ,rust-lang    ,rust       ,240           ,249           ,2021-07-01 ,2021-08-01
  2 ,rust-lang    ,rustup     ,9             ,6             ,2021-07-01 ,2021-08-01
  3 ,rust-lang    ,rustfmt    ,22            ,15            ,2021-07-01 ,2021-08-01
  4 ,rust-lang    ,rustfix    ,0             ,0             ,2021-07-01 ,2021-08-01
  5 ,rust-lang    ,cargo      ,42            ,34            ,2021-07-01 ,2021-08-01
  6 ,rust-lang    ,crates.io  ,7             ,11            ,2021-07-01 ,2021-08-01
  7 ,rust-lang    ,docs.rs    ,7             ,5             ,2021-07-01 ,2021-08-01
  8 ,rust-lang    ,mdBook     ,13            ,14            ,2021-07-01 ,2021-08-01
  ```

We should merge the logic in each metric into `metrics/list_repos.rs` to generate the following CSV in `$DATA_DIR/inputs/repo-list.csv`

```csv
# ,Organization ,Repository ,PRs Opened ,Issues Opened ,Issues Closed ,Start Date ,End Date
1 ,rust-lang    ,rust       ,452        ,240           ,249           ,2021-07-01 ,2021-08-01
2 ,rust-lang    ,rustup     ,11         ,9             ,6             ,2021-07-01 ,2021-08-01
3 ,rust-lang    ,rustfmt    ,12         ,22            ,15            ,2021-07-01 ,2021-08-01
4 ,rust-lang    ,rustfix    ,0          ,0             ,0             ,2021-07-01 ,2021-08-01
5 ,rust-lang    ,cargo      ,38         ,42            ,34            ,2021-07-01 ,2021-08-01
6 ,rust-lang    ,crates.io  ,27         ,7             ,11            ,2021-07-01 ,2021-08-01
7 ,rust-lang    ,docs.rs    ,4          ,7             ,5             ,2021-07-01 ,2021-08-01
8 ,rust-lang    ,mdBook     ,9          ,13            ,14            ,2021-07-01 ,2021-08-01
```

### PR DESCRIPTION
**Pull request by chazkiker2**
### Description

- [x] fix #64 
- [x] combine the `Producer` functionality for `issue_closures` and `list_repos` into one producer
  - [x] fix associated consumers
- [x] add some clarifying comments & doc-comments throughout

r? @rylev

### CODE REVIEW

File: src/report/issue_closure.rs [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 use serde::{Deserialize, Serialize};
 use stable_eyre::eyre::{Error, WrapErr};
 
 use super::{Report, ReportConfig, ReportData};
 
-use std::fs::File;
-use std::path::Path;
-
 #[derive(Debug, Serialize, Deserialize)]
 pub(super) struct IssueClosure {
```
Comment (rylev) [inline src/report/issue_closure.rs:13]: We don't need this struct at all any more. We just need the `write_issue_closures` function which can read info from `ReportData.repo_infos`

**Review by rylev — CHANGES_REQUESTED** [review]
This is definitely a good step in the right direction. I think we don't need there to always be a 1 to 1 relationship between input and output. We just need the repo infos input (and remove the idea of an issue closures input) which the issue closures output uses for the information it needs.

File: src/metrics/list_repos.rs [hunk 2]
```diff
@@ -48,10 +58,66 @@ impl Producer for ListReposForOrg {
...
+/// - `end_date` — The end of the relevant time period to search within
+#[throws]
+async fn count_issue_closures(
+    graphql: &mut Graphql,
+    org_name: &str,
+    repo_name: &str,
+    start_date: &Datetime,
+    end_date: &Datetime,
+) -> IssueClosuresCount {
+    let opened = util::count_issues(
```
[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/metrics/list_repos.rs:108]: I'd be happy for these to run in parellel; 

is this as simple as `tokio::spawn` and `clone` any necessary data?

[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/metrics/list_repos.rs]: gave this a shot... you'll have to let me know if I missed the mark :)

File: src/report/issue_closure.rs [hunk 3]
```diff
@@ -60,4 +61,21 @@ impl Report {
             .unwrap();
         }
     }
+
+    #[throws]
+    pub(super) async fn issue_closures(&self, _config: &ReportConfig) -> Vec<IssueClosure> {
+        debug!("Finding issue closures...");
+        // path to relevant input data
+        let repo_info = self.input_dir().join("repo-infos.csv");
```
[the remark this answers is not in the document]
Comment (chazkiker2) [reply src/report/issue_closure.rs]: agreed that this was not ideal naming. But now that we removed the `IssueClosure` struct, maybe the `repo-infos` name feels more appropriate?

Happy to rename it, too, if you have any ideas.

**Comment by chazkiker2** [comment]
Changes since last review: 

- [x] removed `IssueClosure` struct
- [x] fixed comment nits
- [x] attempted parallelization of the various count queries... but hoping to have some more experienced eyes on that :smile:



### `--replay-graphql` bug
- I found an issue with the `--replay-graphql`... not sure if this is a bug I introduced here or if it existed. 
- still need to dig deeper and figure out why this is happening
- the changes in this branch are working great without the `--replay-graphql` option; but as soon we run it WITH that flag, the GQL query is loading response data from the same file for both `metrics::util::count_pull_requests` and `metrics::util::count_issues`
 - running with `RUST_LOG=debug` demonstrates this issue well

    <details>
    
    <summary>expand to see logging output</summary>
        
    ```bash
    ❯ RUST_LOG=debug cargo run -- report data/optopodi/2021-07-31                 
       Compiling optopodi v0.1.0 (/Users/chazadmin/code/opensource/github-metrics)
        Finished dev [unoptimized + debuginfo] target(s) in 7.84s
         Running `target/debug/optopodi report data/optopodi/2021-07-31`
    [2021-07-31T18:13:32Z DEBUG reqwest::connect] starting new connection: https://api.github.com/
    [2021-07-31T18:13:32Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:33Z DEBUG reqwest::async_impl::client] response '200 OK' for https://api.github.com/graphql
    [2021-07-31T18:13:33Z DEBUG optopodi::metrics::list_repos] Retried issue closure info for optopodi/optopodi: IssueClosuresCount { opened: 2, closed: 5 }
    ```
    
    ```shell
    
    ❯ RUST_LOG=debug cargo run -- --replay-graphql report data/optopodi/2021-07-31
        Finished dev [unoptimized + debuginfo] target(s) in 0.15s
         Running `target/debug/optopodi --replay-graphql report data/optopodi/2021-07-31`
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/all-repos/0.json` rather than github
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-participants/0.json` rather than github
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-infos/0.json` rather than github
    [2021-07-31T18:13:35Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-infos/0.json` rather than github
    [2021-07-31T18:13:35Z DEBUG optopodi::metrics::util] Fetching issue closure info for optopodi/optopodi
    [2021-07-31T18:13:35Z INFO  optopodi::metrics::gql] loading response data from `data/optopodi/2021-07-31/graphql/repo-infos/0.json` rather than github
    [2021-07-31T18:13:35Z DEBUG optopodi::metrics::list_repos] Retried issue closure info for optopodi/optopodi: IssueClosuresCount { opened: 5, closed: 5 }
    ```
        
    </details>

**Comment by chazkiker2** [comment]
is it possible that would happen due to the multi-threaded context?

File: src/metrics/list_repos.rs [hunk 4]
```diff
@@ -32,26 +34,119 @@ impl ListReposForOrg {
     }
 }
 
+impl ListReposForOrg {
+    fn make_repo(&self, repo_name: &str) -> Repo {
```
Comment (rylev) [inline src/metrics/list_repos.rs:38]: I think `to_repo` would be a more appropriate name

File: src/metrics/list_repos.rs [hunk 5]
```diff
@@ -32,26 +34,119 @@ impl ListReposForOrg {
...
+    repo_name: String,
+    start_date: Datetime,
+    end_date: Datetime,
+}
+
+impl Repo {
+    #[throws]
+    async fn spawn_count_pulls(&self) -> usize {
+        let mut repo = self.clone();
+        tokio::spawn(async move { repo.count_pulls().await }).await??
```
Comment (rylev) [inline src/metrics/list_repos.rs:105]: This doesn't really help: we spawn the task and then immediately await the result. This is analogous to spawing and thread and immediately joining on that thread. 

I think we should back this out and just get the pulls count and issue closures count one after the other.

File: src/metrics/list_repos.rs [hunk 6]
```diff
@@ -32,26 +34,119 @@ impl ListReposForOrg {
...
+    }
+
+    #[throws]
+    async fn spawn_count_issue_closures(&self) -> IssueClosuresCount {
+        let mut repo = self.clone();
+        let opened = tokio::spawn(async move { repo.count_issues("created").await }).await??;
+
+        let mut repo = self.clone();
+        let closed = tokio::spawn(async move { repo.count_issues("closed").await }).await??;
```
Comment (rylev) [inline src/metrics/list_repos.rs:114]: As with the above, this still means that the opened and closed counts are retrieved in order. Let's removing the spawning for now.

**Review by rylev — CHANGES_REQUESTED** [review]
It looks like the graphql caching does not take arguments into account so when we make the same request (a GitHub query) with a different query string, it reads the same response even though the request was different (because they query strings were different.

File: src/metrics/list_repos.rs [hunk 7]
```diff
@@ -32,26 +34,116 @@ impl ListReposForOrg {
...
+        let mut clone = self.clone();
+
+        let futures = vec![repo.count_issues("created"), clone.count_issues("closed")];
+
+        let resolved = futures::future::try_join_all(futures).await?;
+
+        let result = IssueClosuresCount {
+            opened: resolved[0],
+            closed: resolved[1],
+        };
```
Comment (chazkiker2) [inline src/metrics/list_repos.rs:113]: @rylev [offered nit](https://github-metrics.zulipchat.com/#narrow/stream/291433-general/topic/issue.20closures.20--.20PR.20.2366/near/247940202):
>  Nit: you can just deconstruct opened and closed in the line above (without needing to first bind to resolved)

Though I had trouble achieving this due to the fact that `resolved` is of type `Vec<usize>` and the compiler refused to deconstruct into a tuple.

```rust
// resolved is type `Vec<usize>`
let resolved = futures::future::try_join_all(futures).await?;
// compiler expected struct `Vec<usize>`, found tuple `(_, _)`
let (opened, closed) = resolved;
```

Would love to include a fix to this nit, if anyone has suggestions.

File: src/metrics/list_repos.rs [hunk 8]
```diff
@@ -32,26 +35,115 @@ impl ListReposForOrg {
...
+    end_date: Datetime,
+}
+
+impl Repo {
+    #[throws]
+    async fn count_issue_closures(&self) -> IssueClosuresCount {
+        let mut repo = self.clone();
+        let mut clone = self.clone();
+
+        let futures = vec![repo.count_issues("created"), clone.count_issues("closed")];
```
Comment (rylev) [inline src/metrics/list_repos.rs:107]: I don't want to delay this any more, but wouldn't [`try_join`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.4/futures/future/trait.TryFutureExt.html#method.try_join) work?

**Review by rylev — APPROVED** [review]
Looks great! 🎉

File: src/metrics/list_repos.rs (hunk 8 above)
Comment (chazkiker2) [reply src/metrics/list_repos.rs]: I do like that better 😄 . pushed up that change

**Pull request merged**